src/ssl/PeekingPeerConnector.cc

```diff
@@ -298,7 +296,7 @@ Ssl::PeekingPeerConnector::noteWantWrite()
     BIO *b = SSL_get_rbio(session.get());
     Ssl::ServerBio *srvBio = static_cast<Ssl::ServerBio *>(BIO_get_data(b));
 
-    if ((srvBio->bumpMode() == Ssl::bumpPeek || srvBio->bumpMode() == Ssl::bumpStare) && srvBio->holdWrite()) {
+    if ((currentBumpMode == Ssl::bumpPeek || currentBumpMode == Ssl::bumpStare) && srvBio->holdWrite()) {
```

Comment:
Is it possible that srvBio->holdWrite() is true when currentBumpMode is _not_ peek or stare? If yes, what kind of situation would lead to that outcome?

I can imagine the following sequence, but currentBumpMode would be peek or stare in this specific case, AFAICT:

1. `ssl_bump peek/stare step2` or similar rule matches.
2. We peek/stare, making holdWrite() true. currentBumpMode is peek/stare here.
3. `ssl_bump splice/bump step3` or similar rule matches.
4. We are trying to wrap things up when TLS triggers noteWantWrite(). But currentBumpMode would still be peek/stare here because it is based on the step2 or earlier decision.

There is identical code in PeekingPeerConnector::noteNegotiationError() as well.

Reply:
> Is it possible that srvBio->holdWrite() is true when currentBumpMode is _not_ peek or stare? If yes, what kind of situation would lead to that outcome?

A bug :-) 

> 
> I can imagine the following sequence, but currentBumpMode would be peek or stare in this specific case, AFAICT:
> 
>     1. `ssl_bump peek/stare step2` or similar rule matches.
> 
>     2. We peek/stare, making holdWrite() true. currentBumpMode is peek/stare here.
> 
>     3. `ssl_bump splice/bump step3` or similar rule matches.

If the decision is bump then the holdWrite is marked false immediately after acl check returns.

If the decision is splice then the holdWrite does not change but we are not going to write or read from SSL any more in this case. 

> 
>     4. We are trying to wrap things up when TLS triggers noteWantWrite(). But currentBumpMode would still be peek/stare here because it is based on the step2 or earlier decision.

The noteWantWrite maybe called if bump was the decision in step3 but in this case the holdWrite is marked false.

The goal of adding currentBumpMode member was to be constant in all PeekingPeerConnector life. Probably it should be renamed.

> 
> 
> There is identical code in PeekingPeerConnector::noteNegotiationError() as well.

Reply:
> > Is it possible that srvBio->holdWrite() is true when currentBumpMode is _not_ peek or stare? If yes, what kind of situation would lead to that outcome?
 
> A bug :-)

If that kind of combination is impossible in the correct code, then we should probable refactor this `if` statement:

```suggestion
    if (srvBio->holdWrite()) {
        // we should not be holding writes unless we are peeking or staring
        Must(currentBumpMode == Ssl::bumpPeek || currentBumpMode == Ssl::bumpStare);
```

The same concern applies to PeekingPeerConnector::noteNegotiationError(), but the correct(ed) code may look different there.